tsnake/initialize.py

```diff
@@ -0,0 +1,543 @@
+        assert p > 0, 'Hyperparameter p (Eq 7) must be > 0 (got p=%f).' % p
+        assert c > 0, 'Hyperparameter c (Eq A.4) must be > 0 (got c=%f).' % c
+        assert a > 0, 'Hyperparameter a (Eqs 1,8) must be > 0 (got a=%f).' % a
+        assert b > 0, 'Hyperparameter b (Eqs 1,8) must be > 0 (got b=%f).' % b
+        assert dt > 0, 'Hyperparameter dt (Eq 8) must be > 0 (got dt=%f).' % dt
+        
+        msg = 'Hyperparameter sigma (Eq A.4) must be > 0 (got sigma=%f).' % sigma
+        assert sigma > 0, msg
+        msg = 'Hyperparameter gamma (Eqs 1,8) must be > 0 (got gamma=%f).' % gamma
+        assert gamma > 0, msg
+        # =============================================================
+        # =============================================================
+        
+        # pixels on the boundary between masked and unmasked
+        edge_pixels = self._find_edge_pixels()
+        
+        step = int(np.floor(len(edge_pixels) / N))
+        step = max(step, 1)
+        if verbose:
+            print('Initializing a Node at every %d-th boundary-pixel' % step)
+        
+        # Order the nodes so that they connect to closest nodes
+        ordered_nodes = self.order_snake_nodes(edge_pixels)
+        
+        # Pull out every step-th pixel for initializion as a Node
+        nodes = []
+        for node_num in range(0, len(ordered_nodes), step):
+            r, c = ordered_nodes[node_num]
+            new_node = Node(r, c)
+            nodes.append(new_node)
+            
+        if verbose:
+            print('Total of %d T-snake nodes were initialized' % len(nodes))
+        
+        force_grid = self.compute_force_grid(sigma, c)
+        intensity_grid = self.raw_image_portion  # grayscale image
+        
+        snake = TSnake(
+            nodes, force_grid, intensity_grid, a, b, gamma, dt
+        )
+        self._initial_tsnake = snake
+        return snake
+    
+    def order_snake_nodes(self, node_locs):
+        """
+        TODO: doc
+        """
+        num_nodes = len(node_locs)
+        node_locs = np.array(node_locs).reshape(num_nodes, 2)
+        
+        # Pull out arbitrary starting node
+        loc = node_locs[0]
+        node_locs = node_locs[1:]
+        
+        ordered_nodes = [loc]
+        while len(ordered_nodes) < num_nodes:
+            # Find the node which is closest to the last node we processed
+            last = ordered_nodes[-1].reshape(1, 2)
+            # dists = distance from each remaining node to current node
+            dists = cdist(last, node_locs).reshape(-1, )
+            
+            closest = dists.argmin()
+            ordered_nodes.append(node_locs[closest])
+            node_locs = np.delete(node_locs, obj=closest, axis=0)
+        
+        return ordered_nodes
+    
+    def compute_force_grid(self, sigma, c):
+        """
+        NOTE: can we manipulate the force grid to make sure the snake
```

Comment:
We can, but in my T-snakes update code, I made sure that all X,Y coordinates remained within the dimensions of the given force grid when updated (which should be the same shape as the image).

Reply:
Yeah I mean something different - if you look at the 'GAN interface demonstration' notebook you might get my idea. It'll be easier to explain what I mean in person, though

(also: note to self to change the name - it only demonstrates the T-snake initialization and doesn't interface with the GAN at all)